File: 2095-delete-the-middle-node-of-a-linked-list/2095-delete-the-middle-node-of-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def lenLinkedList(self, head: Optional[ListNode]) -> int:
        root = head
        count = 0
        while root:
            count += 1
            root = root.next
        return count

    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        k = self.lenLinkedList(head) // 2  # k = n // 2
        if k == 0:
            return None
        root = head
        while k > 1:
            root = root.next
            k -= 1
        root.next = root.next.next
        return head

File: 2095-delete-the-middle-node-of-a-linked-list/2095-delete-the-middle-node-of-a-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int lenLinkedList(ListNode* head) {
        ListNode* root = head;
        int count = 0;
        while (root) {
            count++;
            root = root->next;
        }
        return count;
    }
    ListNode* deleteMiddle(ListNode* head) {
        int k = this->lenLinkedList(head) / 2;  // k = n // 2
        if (k == 0) {
            return NULL;
        }
        ListNode* root = head;
        while (k > 1) {
            root = root->next;
            k--;
        }
        root->next = root->next->next;
        return head;
    }
};